### PR TITLE
SI-6951 Refine ClassTag based unapply after inference.

### PR DESCRIPTION
**Pull request by retronym**
Before, we were getting:

```
typed ClassTag.apply[T#Branch](classOf[Tree$Branch]).unapply(<unapply-selector>) <unapply> (Matcher.unapply[Oak.type](<unapply-selector>) <unapply> ((branch @ _))): T#Branch
```

This is problematic for two reasons:
- We don't actually need to use a class tag in
  pattern matching, because after inference the
  substituted formal type of the unapply method
  is a concrete type, Oak.Branch.
- The class tag check delivers a T#Branch, which
  doesn't line up with the required Oak.Branch
  required by unapply.

Now, we use the inferred type argument `Oak.type` to
double check whether the type is in fact uncheckable,
and retain the inferred type.

In addition, we only consider class tags available
through normal implcits, we disable materialization.
See the code comments for an example of an uncheckable
type that for which a classtag can be materialized.

Review by @adriaanm

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -3478,11 +3479,26 @@ trait Typers extends Modes with Adaptations with Tags {
...
+            }
+            inferredScrutineeType flatMap (extractorForUncheckedType(tree.pos, _))
+          }
+
           // if the type that the unapply method expects for its argument is uncheckable, wrap in classtag extractor
           // skip if the unapply's type is not a method type with (at least, but really it should be exactly) one argument
           // also skip if we already wrapped a classtag extractor (so we don't keep doing that forever)
-          if (uncheckedTypeExtractor.isEmpty || fun1.symbol.owner.isNonBottomSubClass(ClassTagClass)) unapply
-          else wrapClassTagUnapply(unapply, uncheckedTypeExtractor.get, unappType.paramTypes.head)
+          val isClassTagMember = fun1.symbol.owner isNonBottomSubClass ClassTagClass
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3497]: make this an `@inline def` so we only do this check when necessary?

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3497]: Why `@inline` in this case?

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3497]: i don't think making isClassTagMember a method will suddenly make the enclosing method inlineable, so at this point we should probably avoid calling small methods?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -3478,11 +3479,26 @@ trait Typers extends Modes with Adaptations with Tags {
...
           val unapply = UnApply(fun1, args1) setPos tree.pos setType itype
 
+          // SI-6951 Use the inferred type arguments of unapply to recheck and recreate
+          //         the class tag based extractor wrapper; we might have inferred a
+          //         checkable type. If we don't do this, we can lose track of singleton types.
+          val uncheckedTypeExtractor = {
+            val inferredScrutineeType: Option[Type] = unappType.paramTypes.headOption.map { tp =>
+              val applied = treeInfo.dissectApplied(fun1)
+              val unapplyFunTpe = applied.core.tpe
+              unapplyFunTpe.paramTypes.head.subst(unapplyFunTpe.typeParams, applied.targs.map(_.tpe))
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3489]: calling `sym.tpe` considered harmful when `sym` may be a type constructor symbol

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -3437,12 +3437,13 @@ trait Typers extends Modes with Adaptations with Tags {
       val argDummy  = context.owner.newValue(nme.SELECTOR_DUMMY, fun.pos, SYNTHETIC) setInfo pt
       val arg       = Ident(argDummy) setType pt
 
-      val uncheckedTypeExtractor =
-        if (unappType.paramTypes.nonEmpty)
-          extractorForUncheckedType(tree.pos, unappType.paramTypes.head)
-        else None
 
       if (!isApplicableSafe(Nil, unappType, List(pt), WildcardType)) {
+        val uncheckedTypeExtractor =
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3442]: be good to add a comment that this `uncheckedTypeExtractor` should imply the `uncheckedTypeExtractor` on line 3485
(to make the canRemedy sound with respect to what we will actually remedy)

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 1 above)
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3497]: Care to elaborate? Enclosing method here is `doTypedUnapply` which is very large method and I don't see why it should be inlineable.

Calling small methods that do not allocate objects is totally fine and recommended practice. JIT will take care of that very easily.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3497]: exactly, it need not be inlineable so increasing its size is not a problem, thus it's okay to inline this small method -- which is what we did (manually) before this PR

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 3 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3442]: Do you believe that is that actually the case? I couldn't convince myself one way or the other. I see how we can miss an unchecked warning if it isn't. Test case suggestions very welcome...

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 1 above)
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3497]: Ok, now I get it. I'd say that our general rule would be that you force inlining with `@inline` if you have a good reason for that and not that it just doesn't hurt. I don't see it here and that's why I asked.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 3 above)
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3442]: I think it's pretty close as it stands: `Forall x. (if(x.nonEmpty) foo(x)).isEmpty => x.map(f).flatMap(foo).isEmpty` as long as `f` doesn't influence the emptiness of `foo`

**Comment by adriaanm** [comment]
ping @retronym -- I think this PR is almost there: changing `.tpe` to `.typeConstructor` is the only major piece of feedback

**Comment by retronym** [comment]
@adriaanm We're using `Tree#tpe` here, rather than `Symbol#{tpe, tpeHK}`. That's admittedly a bit weird, but was the only way I could figure out to get the inferred type arguments of the `unapply` call. I've added a test to show that is works in the presence of type params of kind `*->*`, and fixed a related bug I find in the process.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 2 above)
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3489]: didn't see it was a tree -- my bad

**Comment by adriaanm** [comment]
closing to keep the queue in check -- please reopen when ready

```
/localhome/jenkins/c/workspace/pr-scala-testsuite-linux-opt/test/files/run/t6951c.scala [FAILED]
```

**Pull request closed without merging**